### ISSUE TITLE
Healthcheck fails in docker when using name instead of IP

### ISSUE THREAD

**Issue by mtrtm** [Xabaril/AspNetCore.Diagnostics.HealthChecks#230]
I have two containers, one that is a net core web api and this one (mssql-docker).

I followed all steps to get the container running.  I can connect using Sql Server Management Studio to the docker container (I mapped 1433 -> 11433 and can use `localhost,11433` to connect).

If I inspect docker networks, I see that the sql server container is on 172.18.0.2.  If I use the following connection string in my web api docker container I can connect to the sql server container: `server=172.18.0.2;Database=master;Integrated Security=False;User ID=sa;Password=abc123ABC123;MultipleActiveResultSets=True`

Obviously the IP will change so I cannot hardcode the IP.  When I change it to use the docker name from my docker-compose, it fails.  Connection string that fails: `"server=sqlserverContainer;Database=master;Integrated Security=False;User ID=sa;Password=abc123ABC123;MultipleActiveResultSets=True"`.  Failure message: `A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)".`

This health check must pass for me to promote this into helm and off my system.  I am using AspNetCore.HealthChecks.SqlServer nuget package and my docker-compose looks like this:

```
sqlserverContainer:
    image: mcr.microsoft.com/mssql/server:2017-latest-ubuntu
    environment:
      - ACCEPT_EULA=Y
      - SA_PASSWORD=abc123ABC123
    ports:
      - 11433:1433
```

In my web api container, I can do the following:
```
 > ping sqlserverContainer
PING sqlserverContainer (172.18.0.2) 56(84) bytes of data.
64 bytes from dockercompose7749112877033456121_sqlserverContainer_1.dockercompose7749112877033456121_default (172.18.0.2): icmp_seq=1 ttl=64 time=0.115 ms
```

I have added a third container running in the same docker network and using the docker-compose name in the connection string works just fine.

Considering I can connect internally (docker network) using the IP, externally using localhost, and internally to a different third mysql container using the docker compose name, I feel this is a config issue with Sql Server or the library that handles the connection string.

Can anyone shed some light why this wouldn't work?

**Comment by CarlosLanderas** [comment]
If you are using this connection string:

`server=sqlserverContainer;Database=master;Integrated Security=False;User ID=sa;Password=abc123ABC123;MultipleActiveResultSets=True`

it will use default sql port that is 1433.

But the port you are binding in your docker-compose is 11433:1433 so it is listening on 11433. I can't see the port specified in the connection string samples

It should be server=sqlserverContainer,11433

**Comment by mtrtm** [comment]
It is not listening on 11433 unless you are talking about anything outside docker.  It proves I have the correct port (default) when the IP address works without another port specification.  The issue here is that, for some reason, the name doesn't work even though it resolves to the correct IP.

**Comment by CarlosLanderas** [comment]
Sorry I missread that they were two containers. Ley me try to reproduce it to check we do not have any regression.  Can you show your health check registration code?  Thanks!

**Comment by CarlosLanderas** [comment]
@mtrtm I have reproduced your problem, it was not able to connect to the sqlserver container, however I have done something unthinkable, that is force the sql  connection string to use port 1433 and it is working now. I do not know if this is something related to the internal working of the Linux Sql Server image but it works:


```
services.AddHealthChecks()
                .AddSqlServer("server=sqlserver,1433, initial catalog=master; user id=sa; password=Password12!", name: "sql1");
```

If I remove the ,1433 (that yes... is the default sql port...) it does not work. It's non sense but is all I can tell you now.

**Comment by mtrtm** [comment]
@CarlosLanderas I really appreciate your attentiveness to this issue and I wish I would have gotten back to you sooner before you closed this.  

I used a connection string with and without 1433 but it still does not work.  

My code to add looks like this:
`return builder.Add(new HealthCheckRegistration(name ?? "sqlserver", (Func<IServiceProvider, IHealthCheck>) (sp => (IHealthCheck) new SqlServerHealthCheck(connectionStringFactory(sp), healthQuery ?? "SELECT 1;")), failureStatus, tags));`

I would share my project but I have a lot of specific work stuff in it.  Do you have a basic one you would like to share on github?  If not I can create a generic one

**Comment by CarlosLanderas** [comment]
@mtrtm,  latest sql server health check supports factory registration if you want to use it. Let me arrive home and I share the project with you.  Are you using windows or Linux sql container?

**Comment by mtrtm** [comment]
linux sql container - trying to keep everything in linux docker so I can 'compose' using docker-compose locally or helm/kubernetes for other environments.

**Comment by mtrtm** [comment]
I have demonstrated this in a very simple solution.  I created an empty web api core project, added docker-compose to it, and set it as startup.  Here it is: https://github.com/mtrtm/CSharp_Aspnetcore_DockerAndSqlServer

**Comment by CarlosLanderas** [comment]
I'm seeing that is the new microsoft mssql recommended image, can you check using this image and tell me if it works for you?. 

microsoft/mssql-server-linux

Maybe it is not working with the new image and I'm gonna check it

**Comment by CarlosLanderas** [comment]
@mtrtm here you have a sample, it's working for me
[](
https://bitbucket.org/clanderas/sample-healthchecks/src/master/)

**Comment by mtrtm** [comment]
@CarlosLanderas - thanks - I cloned and set the startup to docker but no dice.  I get the following when going to localhost:10000/health

{"status":"Unhealthy","totalDuration":"00:00:14.6150029","entries":{"sql1":{"data":{},"description":"A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)","duration":"00:00:14.5954603","exception":"A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)","status":"Unhealthy"}}}

Have you tried mine?  What are we doing differently?  I am on Windows 10 and using VS 2019.

**Comment by CarlosLanderas** [comment]
I dont use vs for docker.  Execute in a cli  docker-compose up -d and see if the same happens

**Comment by mtrtm** [comment]
Wow - this is crazy.  When I do docker-compose with my project I still get unhealthy.  When I do it with yours, I get healthy.  I have no idea what is different between launching your project from VS vs cli, but there is obviously something since VS doesn't work but cli does

**Comment by CarlosLanderas** [comment]
Is is infrastructure related then.  👍

**Comment by CarlosLanderas** [comment]
@mtrtm your problem Is here. You should configure the ports in the override file or remove it.  They are being overriden

https://github.com/mtrtm/CSharp_Aspnetcore_DockerAndSqlServer/blob/master/docker-compose.override.yml

**Comment by mtrtm** [comment]
I have reported this here https://developercommunity.visualstudio.com/content/problem/653229/docker-compose-not-working-with-sql-server.html

Hopefully they will fix it, but I also posted a workaround (don't use capital letters when naming your container)
